File: frontend/src/utils/compositionTimeline.js
/**
 * Variable-tempo / variable-meter timeline helpers for composition.v2.
 * Pure functions derived only from canonical V2 fields. No runtime logging
 * in conversion hot paths; callers log compile summaries.
 */

const SUPPORTED_DENOMINATORS = new Set([1, 2, 4, 8, 16, 32]);

export function parseTimeSignature(timeSignature) {
  if (!timeSignature || typeof timeSignature !== 'string') {
    return null;
  }
  const parts = timeSignature.trim().split('/');
  if (parts.length !== 2) {
    return null;
  }
  const numerator = Number(parts[0]);
  const denominator = Number(parts[1]);
  if (!Number.isInteger(numerator) || !Number.isInteger(denominator)) {
    return null;
  }
  if (numerator < 1 || numerator > 32 || !SUPPORTED_DENOMINATORS.has(denominator)) {
    return null;
  }
  return { numerator, denominator };
}

export function barDurationTicks(timeSignature, ticksPerQuarter) {
  const parsed = parseTimeSignature(timeSignature);
  const tpq = Number(ticksPerQuarter);
  if (!parsed || !Number.isInteger(tpq) || tpq <= 0) {
    return null;
  }
  const numeratorTicks = parsed.numerator * 4 * tpq;
  if (numeratorTicks % parsed.denominator !== 0) {
    return null;
  }
  return numeratorTicks / parsed.denominator;
}

export function roundHalfAwayFromZero(value) {
  if (value >= 0) {
    return Math.trunc(value + 0.5);
  }
  return Math.trunc(value - 0.5);
}

/**
 * @returns {number[]|null} bar start ticks plus final end tick (length barCount+1)
 */
export function compileBarBoundaries({
  timeSignature,
  ticksPerQuarter,
  barCount,
  durationTicks,
  timeSignatureChanges = [],
} = {}) {
  const tpq = Number(ticksPerQuarter);
  const bars = Number(barCount);
  const duration = Number(durationTicks);
  if (!Number.isInteger(tpq) || tpq <= 0 || !Number.isInteger(bars) || bars < 1) {
    return null;
  }
  if (!Number.isInteger(duration) || duration <= 0) {
    return null;
  }

  const changeByTick = new Map();
  for (const change of timeSignatureChanges || []) {
    const tick = Number(change.tick);
    const meter = change.time_signature;
    if (!Number.isInteger(tick) || tick <= 0 || tick >= duration) {
      return null;
    }
    if (changeByTick.has(tick)) {
      return null;
    }
    if (barDurationTicks(meter, tpq) == null) {
      return null;
    }
    changeByTick.set(tick, meter);
  }

  const boundaries = [0];
  let activeMeter = timeSignature;
  for (let index = 0; index < bars; index += 1) {
    const start = boundaries[boundaries.length - 1];
    if (start !== 0 && changeByTick.has(start)) {
      activeMeter = changeByTick.get(start);
    }
    const barTicks = barDurationTicks(activeMeter, tpq);
    if (barTicks == null) {
      return null;
    }
    boundaries.push(start + barTicks);
  }

  if (boundaries[boundaries.length - 1] !== duration) {
    return null;
  }

  for (const tick of changeByTick.keys()) {
    if (!boundaries.slice(0, -1).includes(tick)) {
      return null;
    }
  }
  return boundaries;
}

function sortedChanges(items, valueKey) {
  return [...(items || [])]
    .map((item) => ({ tick: Number(item.tick), value: item[valueKey] }))
    .filter((item) => Number.isInteger(item.tick))
    .sort((a, b) => a.tick - b.tick);
}

function activeFromChanges(rootValue, changes, tick, valueKey = 'value') {
  let current = rootValue;
  for (const change of changes) {
    if (change.tick <= tick) {
      current = change[valueKey];
    } else {
      break;
    }
  }
  return current;
}

/**
 * Compile a timeline object from a composition document.
 * @returns {object|null}
 */
export function compileTimeline(composition) {
  if (!composition || typeof composition !== 'object') {
    return null;
  }
  const ticksPerQuarter = Number(composition.ticks_per_quarter) || 480;
  const durationTicks = Number(composition.duration_ticks);
  const barCount = Number(composition.bar_count);
  const rootTempo = Number(composition.tempo);
  const rootMeter = composition.time_signature;
  const rootKey = composition.key;
  const tempoChanges = sortedChanges(composition.tempo_changes, 'bpm').map((item) => ({
    tick: item.tick,
    value: Number(item.value),
  }));
  const meterChanges = sortedChanges(composition.time_signature_changes, 'time_signature');
  const keyChanges = sortedChanges(composition.key_changes, 'key');

  const barBoundaries = compileBarBoundaries({
    timeSignature: rootMeter,
    ticksPerQuarter,
    barCount,
    durationTicks,
    timeSignatureChanges: composition.time_signature_changes || [],
  });
  if (!barBoundaries) {
    return null;
  }

  return {
    ticksPerQuarter,
    durationTicks,
    barCount,
    barBoundaries,
    rootTempo,
    rootTimeSignature: rootMeter,
    rootKey,
    tempoChanges,
    timeSignatureChanges: meterChanges,
    keyChanges,
  };
}

export function barStartTick(timeline, bar) {
  if (!timeline || bar < 1 || bar > timeline.barCount) {
    return null;
  }
  return timeline.barBoundaries[bar - 1];
}

export function barEndTick(timeline, bar) {
  if (!timeline || bar < 1 || bar > timeline.barCount) {
    return null;
  }
  return timeline.barBoundaries[bar];
}

export function barRangeTicks(timeline, startBar, endBar) {
  if (!timeline || endBar < startBar) {
    return null;
  }
  const startTick = barStartTick(timeline, startBar);
  const endTick = barEndTick(timeline, endBar);
  if (startTick == null || endTick == null) {
    return null;
  }
  return { startTick, endTick };
}

export function barAtTick(timeline, tick) {
  if (!timeline || tick < 0 || tick > timeline.durationTicks) {
    return null;
  }
  if (tick === timeline.durationTicks) {
    return timeline.barCount;
  }
  let lo = 0;
  let hi = timeline.barCount - 1;
  while (lo <= hi) {
    const mid = Math.floor((lo + hi) / 2);
    const start = timeline.barBoundaries[mid];
    const end = timeline.barBoundaries[mid + 1];
    if (tick >= start && tick < end) {
      return mid + 1;
    }
    if (tick < start) {
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }
  return timeline.barCount;
}

export function activeTempo(timeline, tick) {
  return activeFromChanges(timeline.rootTempo, timeline.tempoChanges, tick);
}

export function activeTimeSignature(timeline, tick) {
  return activeFromChanges(timeline.rootTimeSignature, timeline.timeSignatureChanges, tick);
}

export function activeKey(timeline, tick) {
  return activeFromChanges(timeline.rootKey, timeline.keyChanges, tick);
}

function integrateTicks(timeline, startTick, endTick) {
  if (endTick < startTick) {
    throw new Error('endTick must be >= startTick');
  }
  if (startTick === endTick) {
    return 0;
  }
  let total = 0;
  let cursor = startTick;
  let tempo = activeTempo(timeline, startTick);
  const upcoming = timeline.tempoChanges.filter((change) => change.tick > startTick).map((change) => change.tick);
  let changeIndex = 0;
  while (cursor < endTick) {
    const nextBoundary = changeIndex < upcoming.length ? upcoming[changeIndex] : endTick;
    const segmentEnd = Math.min(nextBoundary, endTick);
    const secondsPerTick = 60 / tempo / timeline.ticksPerQuarter;
    total += (segmentEnd - cursor) * secondsPerTick;
    cursor = segmentEnd;
    if (changeIndex < upcoming.length && upcoming[changeIndex] === cursor) {
      tempo = activeTempo(timeline, cursor);
      changeIndex += 1;
    }
  }
  return total;
}

export function tickToSeconds(timeline, tick) {
  if (!timeline || tick < 0 || tick > timeline.durationTicks) {
    return null;
  }
  return integrateTicks(timeline, 0, tick);
}

export function totalDurationSeconds(timeline) {
  if (!timeline) {
    return null;
  }
  return integrateTicks(timeline, 0, timeline.durationTicks);
}

export function secondsToTick(timeline, seconds) {
  if (!timeline || seconds < 0) {
    return null;
  }
  const total = totalDurationSeconds(timeline);
  if (seconds >= total) {
    return timeline.durationTicks;
  }

  let cursorTick = 0;
  let remaining = seconds;
  let tempo = timeline.rootTempo;
  let changeIndex = 0;
  while (cursorTick < timeline.durationTicks) {
    const nextChange = changeIndex < timeline.tempoChanges.length
      ? timeline.tempoChanges[changeIndex].tick
      : timeline.durationTicks;
    const segmentEnd = Math.min(nextChange, timeline.durationTicks);
    if (segmentEnd <= cursorTick) {
      if (changeIndex < timeline.tempoChanges.length && nextChange === cursorTick) {
        tempo = timeline.tempoChanges[changeIndex].value;
        changeIndex += 1;
        continue;
      }
      break;
    }
    const secondsPerTick = 60 / tempo / timeline.ticksPerQuarter;
    const segmentTicks = segmentEnd - cursorTick;
    const segmentSeconds = segmentTicks * secondsPerTick;
    if (remaining <= segmentSeconds) {
      return cursorTick + (remaining / secondsPerTick);
    }
    remaining -= segmentSeconds;
    cursorTick = segmentEnd;
    if (changeIndex < timeline.tempoChanges.length && nextChange === cursorTick) {
      tempo = timeline.tempoChanges[changeIndex].value;
      changeIndex += 1;
    }
  }
  return timeline.durationTicks;
}

/**
 * Map pointer X to a 1-based bar using variable-width bar boundaries.
 */
export function pointerXToBarFromTimeline(pointerX, {
  timeline,
  pixelsPerTick,
  scrollLeft = 0,
} = {}) {
  if (!timeline || !Number.isFinite(Number(pixelsPerTick)) || Number(pixelsPerTick) <= 0) {
    return { bar: null, warning: 'timeline and pixelsPerTick are required' };
  }
  const x = Number(pointerX) + Number(scrollLeft || 0);
  if (!Number.isFinite(x)) {
    return { bar: null, warning: 'pointerX must be a number' };
  }
  const tick = Math.max(0, Math.min(timeline.durationTicks, x / Number(pixelsPerTick)));
  const bar = barAtTick(timeline, tick);
  return { bar };
}

File: backend/app/services/composition_timeline.py
"""Variable-tempo / variable-meter timeline compilation for composition.v2."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any, Sequence

from app.composition_schemas import compile_bar_boundaries, round_half_away_from_zero
from app.services.composition_timing import bar_duration_ticks


logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class TimelineState:
    tick: int
    tempo: int
    time_signature: str
    key: str
    bar: int


@dataclass(frozen=True)
class CompiledTimeline:
    """Compiled bar/tempo/key map derived only from canonical composition fields."""

    ticks_per_quarter: int
    duration_ticks: int
    bar_count: int
    bar_boundaries: tuple[int, ...]  # length bar_count + 1
    root_tempo: int
    root_time_signature: str
    root_key: str
    tempo_changes: tuple[tuple[int, int], ...]  # (tick, bpm)
    time_signature_changes: tuple[tuple[int, str], ...]
    key_changes: tuple[tuple[int, str], ...]

    def bar_start_tick(self, bar: int) -> int:
        if bar < 1 or bar > self.bar_count:
            raise ValueError(f"bar {bar} is outside 1..{self.bar_count}")
        return self.bar_boundaries[bar - 1]

    def bar_end_tick(self, bar: int) -> int:
        if bar < 1 or bar > self.bar_count:
            raise ValueError(f"bar {bar} is outside 1..{self.bar_count}")
        return self.bar_boundaries[bar]

    def bar_range_ticks(self, start_bar: int, end_bar: int) -> tuple[int, int]:
        """Inclusive start_bar / end_bar → [start_tick, end_tick)."""
        if end_bar < start_bar:
            raise ValueError("end_bar must be >= start_bar")
        return self.bar_start_tick(start_bar), self.bar_end_tick(end_bar)

    def bar_at_tick(self, tick: int) -> int:
        if tick < 0 or tick > self.duration_ticks:
            raise ValueError("tick is outside composition duration")
        if tick == self.duration_ticks:
            return self.bar_count
        # Find largest bar start <= tick
        lo, hi = 0, self.bar_count - 1
        while lo <= hi:
            mid = (lo + hi) // 2
            start = self.bar_boundaries[mid]
            end = self.bar_boundaries[mid + 1]
            if start <= tick < end:
                return mid + 1
            if tick < start:
                hi = mid - 1
            else:
                lo = mid + 1
        return self.bar_count

    def active_tempo(self, tick: int) -> int:
        tempo = self.root_tempo
        for change_tick, bpm in self.tempo_changes:
            if change_tick <= tick:
                tempo = bpm
            else:
                break
        return tempo

    def active_time_signature(self, tick: int) -> str:
        meter = self.root_time_signature
        for change_tick, signature in self.time_signature_changes:
            if change_tick <= tick:
                meter = signature
            else:
                break
        return meter

    def active_key(self, tick: int) -> str:
        key = self.root_key
        for change_tick, value in self.key_changes:
            if change_tick <= tick:
                key = value
            else:
                break
        return key

    def state_at_tick(self, tick: int) -> TimelineState:
        return TimelineState(
            tick=tick,
            tempo=self.active_tempo(tick),
            time_signature=self.active_time_signature(tick),
            key=self.active_key(tick),
            bar=self.bar_at_tick(tick) if tick <= self.duration_ticks else self.bar_count,
        )

    def tick_to_seconds(self, tick: int) -> float:
        if tick < 0:
            raise ValueError("tick cannot be negative")
        if tick > self.duration_ticks:
            raise ValueError("tick exceeds composition duration")
        return self._integrate_ticks(0, tick)

    def seconds_to_tick(self, seconds: float) -> float:
        if seconds < 0:
            raise ValueError("seconds cannot be negative")
        total = self.total_duration_seconds()
        if seconds >= total:
            return float(self.duration_ticks)

        # Walk tempo segments until cumulative seconds reach target.
        cursor_tick = 0
        remaining = float(seconds)
        tempo = self.root_tempo
        change_index = 0
        while cursor_tick < self.duration_ticks:
            next_change = (
                self.tempo_changes[change_index][0]
                if change_index < len(self.tempo_changes)
                else self.duration_ticks
            )
            segment_end = min(next_change, self.duration_ticks)
            if segment_end <= cursor_tick:
                if change_index < len(self.tempo_changes) and next_change == cursor_tick:
                    tempo = self.tempo_changes[change_index][1]
                    change_index += 1
                    continue
                break
            seconds_per_tick = 60.0 / tempo / self.ticks_per_quarter
            segment_ticks = segment_end - cursor_tick
            segment_seconds = segment_ticks * seconds_per_tick
            if remaining <= segment_seconds:
                return cursor_tick + (remaining / seconds_per_tick)
            remaining -= segment_seconds
            cursor_tick = segment_end
            if change_index < len(self.tempo_changes) and next_change == cursor_tick:
                tempo = self.tempo_changes[change_index][1]
                change_index += 1
        return float(self.duration_ticks)

    def total_duration_seconds(self) -> float:
        return self._integrate_ticks(0, self.duration_ticks)

    def _integrate_ticks(self, start_tick: int, end_tick: int) -> float:
        if end_tick < start_tick:
            raise ValueError("end_tick must be >= start_tick")
        if start_tick == end_tick:
            return 0.0

        total = 0.0
        cursor = start_tick
        tempo = self.active_tempo(start_tick)
        # Upcoming tempo change ticks strictly after start.
        upcoming = [tick for tick, _bpm in self.tempo_changes if tick > start_tick]
        change_index = 0
        while cursor < end_tick:
            next_boundary = upcoming[change_index] if change_index < len(upcoming) else end_tick
            segment_end = min(next_boundary, end_tick)
            seconds_per_tick = 60.0 / tempo / self.ticks_per_quarter
            total += (segment_end - cursor) * seconds_per_tick
            cursor = segment_end
            if change_index < len(upcoming) and upcoming[change_index] == cursor:
                tempo = self.active_tempo(cursor)
                change_index += 1
        return total

    def summary_extra(self) -> dict[str, Any]:
        return {
            "bar_count": self.bar_count,
            "duration_ticks": self.duration_ticks,
            "tempo_segment_count": len(self.tempo_changes) + 1,
            "meter_change_count": len(self.time_signature_changes),
            "key_change_count": len(self.key_changes),
            "total_seconds": round(self.total_duration_seconds(), 6),
        }


def _change_pairs(items: Sequence[Any], tick_attr: str, value_attr: str) -> tuple[tuple[int, Any], ...]:
    pairs: list[tuple[int, Any]] = []
    for item in items or ():
        if hasattr(item, tick_attr):
            tick = int(getattr(item, tick_attr))
            value = getattr(item, value_attr)
        else:
            tick = int(item[tick_attr])
            value = item[value_attr]
        pairs.append((tick, value))
    return tuple(sorted(pairs, key=lambda pair: pair[0]))


def compile_timeline(composition: Any) -> CompiledTimeline:
    """Compile a timeline from a CompositionV2-like object or mapping."""
    if hasattr(composition, "model_dump"):
        data = composition
        ticks_per_quarter = int(data.ticks_per_quarter)
        duration_ticks = int(data.duration_ticks)
        bar_count = int(data.bar_count)
        root_tempo = int(data.tempo)
        root_meter = str(data.time_signature)
        root_key = str(data.key)
        tempo_changes = getattr(data, "tempo_changes", ()) or ()
        meter_changes = getattr(data, "time_signature_changes", ()) or ()
        key_changes = getattr(data, "key_changes", ()) or ()
    else:
        data = composition
        ticks_per_quarter = int(data.get("ticks_per_quarter") or 480)
        duration_ticks = int(data["duration_ticks"])
        bar_count = int(data["bar_count"])
        root_tempo = int(data["tempo"])
        root_meter = str(data["time_signature"])
        root_key = str(data["key"])
        tempo_changes = data.get("tempo_changes") or ()
        meter_changes = data.get("time_signature_changes") or ()
        key_changes = data.get("key_changes") or ()

    boundaries = compile_bar_boundaries(
        time_signature=root_meter,
        ticks_per_quarter=ticks_per_quarter,
        bar_count=bar_count,
        duration_ticks=duration_ticks,
        time_signature_changes=list(meter_changes),
    )
    timeline = CompiledTimeline(
        ticks_per_quarter=ticks_per_quarter,
        duration_ticks=duration_ticks,
        bar_count=bar_count,
        bar_boundaries=tuple(boundaries),
        root_tempo=root_tempo,
        root_time_signature=root_meter,
        root_key=root_key,
        tempo_changes=_change_pairs(tempo_changes, "tick", "bpm"),
        time_signature_changes=_change_pairs(meter_changes, "tick", "time_signature"),
        key_changes=_change_pairs(key_changes, "tick", "key"),
    )
    logger.info(
        "Compiled composition timeline",
        extra=timeline.summary_extra(),
    )
    logger.debug(
        "Timeline compile details",
        extra={
            "boundary_count": len(timeline.bar_boundaries),
            "first_bar_ticks": timeline.bar_boundaries[1] - timeline.bar_boundaries[0] if bar_count else 0,
            "root_tempo": root_tempo,
            "root_time_signature": root_meter,
        },
    )
    return timeline


def constant_bar_ticks(time_signature: str, ticks_per_quarter: int) -> int:
    """Compatibility helper for single-meter consumers."""
    return bar_duration_ticks(time_signature, ticks_per_quarter)


def round_tick(value: float) -> int:
    return round_half_away_from_zero(value)
